File: 03_use_cases/13_macos_cua/python/metacua/overlay.py
"""Transparent, click-through cursor overlay driven with CoreAnimation.

Renders a small glowing animated cursor at each action the agent performs so the
user can watch the model drive the machine. All public methods are thread-safe
(they hop to the main thread); the agent loop calls them from its background thread.
"""

import objc
from AppKit import (
    NSBackingStoreBuffered,
    NSColor,
    NSView,
    NSWindow,
    NSWindowStyleMaskBorderless,
)
from Foundation import NSMakeRect, NSMakePoint, NSObject
from Quartz import (
    CABasicAnimation,
    CAKeyframeAnimation,
    CALayer,
    CAMediaTimingFunction,
    CAShapeLayer,
    CATextLayer,
    CATransaction,
    CGPathCreateWithEllipseInRect,
    CGPointMake,
    CGRectMake,
    CGShieldingWindowLevel,
    kCAMediaTimingFunctionEaseInEaseOut,
    kCAMediaTimingFunctionEaseOut,
)
import Quartz

from .screenshot import primary_screen

# NSWindow collection-behavior flags (not all are exported by name in pyobjc).
_CAN_JOIN_ALL_SPACES = 1 << 0
_STATIONARY = 1 << 4
_FULLSCREEN_AUXILIARY = 1 << 8
_IGNORES_CYCLE = 1 << 6


def dispatch_main(work):
    """Run `work` on the main thread asynchronously."""
    from Foundation import NSOperationQueue

    NSOperationQueue.mainQueue().addOperationWithBlock_(work)


class OverlayController:
    # A warm rose accent (not blue), with a soft glow.
    _accent = NSColor.colorWithCalibratedRed_green_blue_alpha_(1.0, 0.36, 0.52, 1.0)
    _radius = 13.0  # smaller cursor
    _hold_seconds = 1.1  # linger after the last action
    _fade_seconds = 0.45

    def __init__(self):
        self._window = None
        self._host_layer = None
        self._cursor_layer = None
        self._label_container = None
        self._label_text = None
        self._screen_height = 0.0
        self._hide_timer = None

        screen = primary_screen()
        if screen is None:
            return
        frame = screen.frame()
        self._screen_height = frame.size.height

        window = NSWindow.alloc().initWithContentRect_styleMask_backing_defer_(
            frame, NSWindowStyleMaskBorderless, NSBackingStoreBuffered, False
        )
        window.setOpaque_(False)
        window.setBackgroundColor_(NSColor.clearColor())
        window.setHasShadow_(False)
        window.setIgnoresMouseEvents_(True)
        window.setLevel_(int(CGShieldingWindowLevel()))
        window.setCollectionBehavior_(
            _CAN_JOIN_ALL_SPACES | _STATIONARY | _FULLSCREEN_AUXILIARY | _IGNORES_CYCLE
        )

        view = NSView.alloc().initWithFrame_(
            NSMakeRect(0, 0, frame.size.width, frame.size.height)
        )
        view.setWantsLayer_(True)
        host = CALayer.layer()
        host.setFrame_(view.bounds())
        view.setLayer_(host)
        window.setContentView_(view)

        scale = screen.backingScaleFactor()

        # Cursor group: glowing ring + center dot, anchored at its center.
        cursor = CALayer.layer()
        d = self._radius * 2
        cursor.setBounds_(CGRectMake(0, 0, d, d))
        cursor.setOpacity_(0)  # hidden until the first action

        ring = CAShapeLayer.layer()
        ring.setPath_(CGPathCreateWithEllipseInRect(CGRectMake(1.5, 1.5, d - 3, d - 3), None))
        ring.setFillColor_(self._accent.colorWithAlphaComponent_(0.12).CGColor())
        ring.setStrokeColor_(self._accent.CGColor())
        ring.setLineWidth_(2)
        ring.setShadowColor_(self._accent.CGColor())
        ring.setShadowRadius_(7)
        ring.setShadowOpacity_(0.85)
        ring.setShadowOffset_(Quartz.CGSizeMake(0, 0))
        ring.setFrame_(cursor.bounds())
        cursor.addSublayer_(ring)

        dot = CAShapeLayer.layer()
        dot_r = 3.0
        dot.setPath_(
            CGPathCreateWithEllipseInRect(
                CGRectMake(self._radius - dot_r, self._radius - dot_r, dot_r * 2, dot_r * 2), None
            )
        )
        dot.setFillColor_(self._accent.CGColor())
        dot.setFrame_(cursor.bounds())
        cursor.addSublayer_(dot)

        host.addSublayer_(cursor)

        # Action label pill below the cursor.
        pill = CALayer.layer()
        pill.setBackgroundColor_(NSColor.blackColor().colorWithAlphaComponent_(0.68).CGColor())
        pill.setCornerRadius_(8)
        pill.setOpacity_(0)
        text = CATextLayer.layer()
        text.setFontSize_(11)
        text.setForegroundColor_(NSColor.whiteColor().CGColor())
        text.setAlignmentMode_("center")
        text.setTruncationMode_("end")
        text.setContentsScale_(scale)
        pill.addSublayer_(text)
        host.addSublayer_(pill)

        self._window = window
        self._host_layer = host
        self._cursor_layer = cursor
        self._label_container = pill
        self._label_text = text

        window.orderFrontRegardless()

    # MARK: - Public API (global top-left coordinates)

    def show_action(self, label, point):
        def work():
            self._place_cursor(point, True, label)
            self._schedule_hide()

        self._on_main(work)

    def show_click(self, label, point):
        def work():
            self._place_cursor(point, True, label)
            self._ripple(point)
            self._bounce()
            self._schedule_hide()

        self._on_main(work)

    def show_drag(self, label, from_point, to_point, duration):
        def work():
            self._place_cursor(from_point, True, label)
            self._ripple(from_point)
            self._place_cursor(to_point, True, label, max(0.2, duration))
            self._schedule_hide()

        self._on_main(work)

    # MARK: - Internals (main thread only)

    def _place_cursor(self, point, animated, label, duration=0.22):
        cursor = self._cursor_layer
        if cursor is None:
            return
        self._cancel_hide()
        p = self._to_view(point)

        CATransaction.begin()
        CATransaction.setAnimationDuration_(duration if animated else 0)
        CATransaction.setAnimationTimingFunction_(
            CAMediaTimingFunction.functionWithName_(kCAMediaTimingFunctionEaseInEaseOut)
        )
        cursor.setPosition_(p)
        cursor.setOpacity_(1)
        self._position_label(p)
        CATransaction.commit()

        self._set_label(label)

    def _position_label(self, p):
        pill = self._label_container
        text = self._label_text
        if pill is None or text is None:
            return
        width = 170.0
        height = 18.0
        pill.setBounds_(CGRectMake(0, 0, width, height))
        pill.setPosition_(CGPointMake(p.x, p.y - self._radius - 15))
        text.setFrame_(CGRectMake(7, 2, width - 14, height - 4))

    def _set_label(self, label):
        pill = self._label_container
        text = self._label_text
        if pill is None or text is None:
            return
        text.setString_(label)
        pill.setOpacity_(0 if not label else 1)

    def _ripple(self, point):
        host = self._host_layer
        if host is None:
            return
        p = self._to_view(point)
        d = self._radius * 2.2
        ripple = CAShapeLayer.layer()
        ripple.setPath_(CGPathCreateWithEllipseInRect(CGRectMake(0, 0, d, d), None))
        ripple.setBounds_(CGRectMake(0, 0, d, d))
        ripple.setPosition_(p)
        ripple.setFillColor_(NSColor.clearColor().CGColor())
        ripple.setStrokeColor_(self._accent.CGColor())
        ripple.setLineWidth_(2)
        host.addSublayer_(ripple)

        scale = CABasicAnimation.animationWithKeyPath_("transform.scale")
        scale.setFromValue_(0.4)
        scale.setToValue_(1.7)
        fade = CABasicAnimation.animationWithKeyPath_("opacity")
        fade.setFromValue_(0.85)
        fade.setToValue_(0)
        group = Quartz.CAAnimationGroup.animation()
        group.setAnimations_([scale, fade])
        group.setDuration_(0.5)
        group.setTimingFunction_(CAMediaTimingFunction.functionWithName_(kCAMediaTimingFunctionEaseOut))
        group.setRemovedOnCompletion_(True)
        ripple.setOpacity_(0)
        ripple.addAnimation_forKey_(group, "ripple")

        self._after(0.55, lambda: ripple.removeFromSuperlayer())

    def _bounce(self):
        cursor = self._cursor_layer
        if cursor is None:
            return
        bounce = CAKeyframeAnimation.animationWithKeyPath_("transform.scale")
        bounce.setValues_([1.0, 0.82, 1.0])
        bounce.setKeyTimes_([0, 0.4, 1.0])
        bounce.setDuration_(0.26)
        bounce.setTimingFunction_(CAMediaTimingFunction.functionWithName_(kCAMediaTimingFunctionEaseOut))
        cursor.addAnimation_forKey_(bounce, "bounce")

    def _cancel_hide(self):
        if self._hide_timer is not None:
            self._hide_timer.invalidate()
            self._hide_timer = None
        if self._cursor_layer is not None:
            self._cursor_layer.removeAnimationForKey_("hide")
        if self._label_container is not None:
            self._label_container.removeAnimationForKey_("hide")

    def _schedule_hide(self):
        if self._hide_timer is not None:
            self._hide_timer.invalidate()
        self._hide_timer = self._after(self._hold_seconds, self._fade_out)

    def _fade_out(self):
        for layer in (self._cursor_layer, self._label_container):
            if layer is None:
                continue
            fade = CABasicAnimation.animationWithKeyPath_("opacity")
            presentation = layer.presentationLayer()
            fade.setFromValue_(presentation.opacity() if presentation is not None else layer.opacity())
            fade.setToValue_(0)
            fade.setDuration_(self._fade_seconds)
            fade.setTimingFunction_(CAMediaTimingFunction.functionWithName_(kCAMediaTimingFunctionEaseOut))
            fade.setFillMode_("forwards")
            fade.setRemovedOnCompletion_(False)
            layer.addAnimation_forKey_(fade, "hide")
            layer.setOpacity_(0)

    def _to_view(self, p):
        return CGPointMake(p.x, self._screen_height - p.y)

    def _on_main(self, work):
        from Foundation import NSThread

        if NSThread.isMainThread():
            work()
        else:
            dispatch_main(work)

    def _after(self, seconds, work):
        """Schedule `work` on the main run loop after `seconds`; returns an NSTimer."""
        from Foundation import NSTimer

        holder = {}

        def handler(timer):
            work()

        timer = NSTimer.timerWithTimeInterval_repeats_block_(seconds, False, handler)
        from Foundation import NSRunLoop, NSDefaultRunLoopMode

        NSRunLoop.mainRunLoop().addTimer_forMode_(timer, NSDefaultRunLoopMode)
        return timer

File: 03_use_cases/13_macos_cua/src/metacua/Overlay.swift
import AppKit
import QuartzCore

/// A transparent, click-through, always-on-top overlay that renders a small
/// glowing animated cursor at each action the agent performs, so the user can
/// watch the LLM drive the machine. After each action the cursor lingers briefly
/// and then fades out. All public methods are thread-safe (they hop to the main
/// thread); the agent loop calls them from its background thread.
final class OverlayController {
  // A warm rose accent (not blue), with a soft glow.
  private let accent = NSColor(calibratedRed: 1.0, green: 0.36, blue: 0.52, alpha: 1.0)
  private let radius: CGFloat = 13 // smaller cursor
  private let holdSeconds = 1.1 // linger after the last action
  private let fadeSeconds = 0.45

  private var window: NSWindow?
  private var hostLayer: CALayer?
  private var cursorLayer: CALayer?
  private var labelContainer: CALayer?
  private var labelText: CATextLayer?
  private var statusContainer: CALayer?
  private var statusTitle: CATextLayer?
  private var statusBody: CATextLayer?
  private var screenHeight: CGFloat = 0
  private var hideWork: DispatchWorkItem?

  /// Build the overlay window. Must be called on the main thread.
  init() {
    guard let screen = primaryScreen() else { return }
    screenHeight = screen.frame.height

    let window = NSWindow(
      contentRect: screen.frame,
      styleMask: .borderless,
      backing: .buffered,
      defer: false
    )
    window.isOpaque = false
    window.backgroundColor = .clear
    window.hasShadow = false
    window.ignoresMouseEvents = true
    window.level = NSWindow.Level(rawValue: Int(CGShieldingWindowLevel()))
    window.collectionBehavior = [
      .canJoinAllSpaces, .stationary, .fullScreenAuxiliary, .ignoresCycle,
    ]

    let view = NSView(frame: NSRect(origin: .zero, size: screen.frame.size))
    view.wantsLayer = true
    let host = CALayer()
    host.frame = view.bounds
    view.layer = host
    window.contentView = view

    let scale = screen.backingScaleFactor

    // Cursor group: glowing ring + center dot, anchored at its center.
    let cursor = CALayer()
    let d = radius * 2
    cursor.bounds = CGRect(x: 0, y: 0, width: d, height: d)
    cursor.opacity = 0 // hidden until the first action

    let ring = CAShapeLayer()
    ring.path = CGPath(
      ellipseIn: CGRect(x: 1.5, y: 1.5, width: d - 3, height: d - 3), transform: nil)
    ring.fillColor = accent.withAlphaComponent(0.12).cgColor
    ring.strokeColor = accent.cgColor
    ring.lineWidth = 2
    ring.shadowColor = accent.cgColor
    ring.shadowRadius = 7
    ring.shadowOpacity = 0.85
    ring.shadowOffset = .zero
    ring.frame = cursor.bounds
    cursor.addSublayer(ring)

    let dot = CAShapeLayer()
    let dotR: CGFloat = 3
    dot.path = CGPath(
      ellipseIn: CGRect(
        x: radius - dotR, y: radius - dotR,
        width: dotR * 2, height: dotR * 2), transform: nil)
    dot.fillColor = accent.cgColor
    dot.frame = cursor.bounds
    cursor.addSublayer(dot)

    host.addSublayer(cursor)

    // Action label pill below the cursor.
    let pill = CALayer()
    pill.backgroundColor = NSColor.black.withAlphaComponent(0.68).cgColor
    pill.cornerRadius = 8
    pill.opacity = 0
    let text = CATextLayer()
    text.fontSize = 11
    text.foregroundColor = NSColor.white.cgColor
    text.alignmentMode = .center
    text.truncationMode = .end
    text.contentsScale = scale
    pill.addSublayer(text)
    host.addSublayer(pill)

    let status = CALayer()
    status.backgroundColor = NSColor.black.withAlphaComponent(0.72).cgColor
    status.cornerRadius = 10
    status.opacity = 0
    status.shadowColor = NSColor.black.cgColor
    status.shadowOpacity = 0.28
    status.shadowRadius = 12
    status.shadowOffset = CGSize(width: 0, height: -2)

    let title = CATextLayer()
    title.fontSize = 12
    title.foregroundColor = accent.cgColor
    title.alignmentMode = .left
    title.truncationMode = .end
    title.contentsScale = scale
    status.addSublayer(title)

    let body = CATextLayer()
    body.fontSize = 12
    body.foregroundColor = NSColor.white.cgColor
    body.alignmentMode = .left
    body.truncationMode = .end
    body.isWrapped = true
    body.contentsScale = scale
    status.addSublayer(body)
    host.addSublayer(status)

    self.window = window
    self.hostLayer = host
    self.cursorLayer = cursor
    self.labelContainer = pill
    self.labelText = text
    self.statusContainer = status
    self.statusTitle = title
    self.statusBody = body

    window.orderFrontRegardless()
  }

  // MARK: - Public API (global top-left coordinates)

  /// Move the cursor to a point and set the action label (no click ripple).
  func showAction(_ label: String, at point: CGPoint) {
    onMain {
      self.placeCursor(at: point, animated: true, label: label)
      self.scheduleHide()
    }
  }

  /// Move the cursor to a point, set the label, and play a click ripple.
  func showClick(_ label: String, at point: CGPoint) {
    onMain {
      self.placeCursor(at: point, animated: true, label: label)
      self.ripple(at: point)
      self.bounce()
      self.scheduleHide()
    }
  }

  /// Animate a drag from one point to another over `duration` seconds.
  func showDrag(_ label: String, from: CGPoint, to: CGPoint, duration: Double) {
    onMain {
      self.placeCursor(at: from, animated: true, label: label)
      self.ripple(at: from)
      self.placeCursor(at: to, animated: true, label: label, duration: max(0.2, duration))
      self.scheduleHide()
    }
  }

  /// Show a fixed, click-through status panel so progress remains visible when Terminal is hidden.
  func showStatus(title: String, body: String) {
    onMain {
      self.setStatus(title: title, body: body)
    }
  }

  func hideStatus() {
    onMain {
      self.statusContainer?.opacity = 0
    }
  }

  // MARK: - Internals (main thread only)

  private func placeCursor(
    at point: CGPoint, animated: Bool, label: String, duration: Double = 0.22
  ) {
    guard let cursor = cursorLayer else { return }
    cancelHide()
    let p = toView(point)

    CATransaction.begin()
    CATransaction.setAnimationDuration(animated ? duration : 0)
    CATransaction.setAnimationTimingFunction(CAMediaTimingFunction(name: .easeInEaseOut))
    cursor.position = p
    cursor.opacity = 1
    positionLabel(under: p)
    CATransaction.commit()

    setLabel(label)
  }

  private func positionLabel(under p: CGPoint) {
    guard let pill = labelContainer, let text = labelText else { return }
    let width: CGFloat = 170
    let height: CGFloat = 18
    pill.bounds = CGRect(x: 0, y: 0, width: width, height: height)
    pill.position = CGPoint(x: p.x, y: p.y - radius - 15)
    text.frame = CGRect(x: 7, y: 2, width: width - 14, height: height - 4)
  }

  private func setLabel(_ label: String) {
    guard let pill = labelContainer, let text = labelText else { return }
    text.string = label
    pill.opacity = label.isEmpty ? 0 : 1
  }

  private func setStatus(title: String, body: String) {
    guard let panel = statusContainer, let titleLayer = statusTitle, let bodyLayer = statusBody,
      let host = hostLayer
    else {
      return
    }
    let margin: CGFloat = 18
    let width = min(CGFloat(420), max(CGFloat(280), host.bounds.width - margin * 2))
    let height: CGFloat = 108
    panel.bounds = CGRect(x: 0, y: 0, width: width, height: height)
    panel.position = CGPoint(x: margin + width / 2, y: host.bounds.height - margin - height / 2)
    titleLayer.frame = CGRect(x: 14, y: height - 30, width: width - 28, height: 18)
    bodyLayer.frame = CGRect(x: 14, y: 12, width: width - 28, height: height - 44)
    titleLayer.string = title
    bodyLayer.string = body
    panel.opacity = 1
  }

  private func ripple(at point: CGPoint) {
    guard let host = hostLayer else { return }
    let p = toView(point)
    let d = radius * 2.2
    let ripple = CAShapeLayer()
    ripple.path = CGPath(ellipseIn: CGRect(x: 0, y: 0, width: d, height: d), transform: nil)
    ripple.bounds = CGRect(x: 0, y: 0, width: d, height: d)
    ripple.position = p
    ripple.fillColor = NSColor.clear.cgColor
    ripple.strokeColor = accent.cgColor
    ripple.lineWidth = 2
    host.addSublayer(ripple)

    let scale = CABasicAnimation(keyPath: "transform.scale")
    scale.fromValue = 0.4
    scale.toValue = 1.7
    let fade = CABasicAnimation(keyPath: "opacity")
    fade.fromValue = 0.85
    fade.toValue = 0
    let group = CAAnimationGroup()
    group.animations = [scale, fade]
    group.duration = 0.5
    group.timingFunction = CAMediaTimingFunction(name: .easeOut)
    group.isRemovedOnCompletion = true
    ripple.opacity = 0
    ripple.add(group, forKey: "ripple")

    DispatchQueue.main.asyncAfter(deadline: .now() + 0.55) { ripple.removeFromSuperlayer() }
  }

  private func bounce() {
    guard let cursor = cursorLayer else { return }
    let bounce = CAKeyframeAnimation(keyPath: "transform.scale")
    bounce.values = [1.0, 0.82, 1.0]
    bounce.keyTimes = [0, 0.4, 1.0]
    bounce.duration = 0.26
    bounce.timingFunction = CAMediaTimingFunction(name: .easeOut)
    cursor.add(bounce, forKey: "bounce")
  }

  /// Cancel any pending fade and make the cursor/label fully visible.
  private func cancelHide() {
    hideWork?.cancel()
    hideWork = nil
    cursorLayer?.removeAnimation(forKey: "hide")
    labelContainer?.removeAnimation(forKey: "hide")
  }

  /// After the linger interval, fade the cursor and label out.
  private func scheduleHide() {
    hideWork?.cancel()
    let work = DispatchWorkItem { [weak self] in self?.fadeOut() }
    hideWork = work
    DispatchQueue.main.asyncAfter(deadline: .now() + holdSeconds, execute: work)
  }

  private func fadeOut() {
    for layer in [cursorLayer, labelContainer].compactMap({ $0 }) {
      let fade = CABasicAnimation(keyPath: "opacity")
      fade.fromValue = layer.presentation()?.opacity ?? layer.opacity
      fade.toValue = 0
      fade.duration = fadeSeconds
      fade.timingFunction = CAMediaTimingFunction(name: .easeOut)
      fade.fillMode = .forwards
      fade.isRemovedOnCompletion = false
      layer.add(fade, forKey: "hide")
      layer.opacity = 0
    }
  }

  private func toView(_ p: CGPoint) -> CGPoint {
    CGPoint(x: p.x, y: screenHeight - p.y)
  }

  private func onMain(_ work: @escaping () -> Void) {
    if Thread.isMainThread {
      work()
    } else {
      DispatchQueue.main.async(execute: work)
    }
  }
}
